### PR TITLE
extend light client protocol for Electra

### PR DESCRIPTION
**Pull request by etan-status**
Add missing Electra support for light client protocol:

- https://github.com/ethereum/consensus-specs/pull/3811

Tested against PR consensus-spec-tests, the test runner automatically picks up the new tests once available.

### CODE REVIEW

File: beacon_chain/spec/forks_light_client.nim [hunk 1]
```diff
@@ -188,8 +213,71 @@ template kind*(
       deneb.LightClientStore]): LightClientDataFork =
   LightClientDataFork.Deneb
 
+template kind*(
+    x: typedesc[ # `SomeLightClientObject` doesn't work here (Nim 1.6)
```
Comment (tersec) [inline beacon_chain/spec/forks_light_client.nim]: Might link to https://github.com/nim-lang/Nim/issues/18095 here

File: beacon_chain/spec/forks_light_client.nim [hunk 2]
```diff
@@ -951,6 +1135,108 @@ func toDenebLightClientHeader(
     execution_branch: blck.message.body.build_proof(
       capella.EXECUTION_PAYLOAD_GINDEX).get)
 
+# https://github.com/ethereum/consensus-specs/blob/v1.5.0-alpha.3/specs/electra/light-client/full-node.md#modified-block_to_light_client_header
+func toElectraLightClientHeader(
+    blck:  # `SomeSignedBeaconBlock` doesn't work here (Nim 1.6)
```
Comment (tersec) [inline beacon_chain/spec/forks_light_client.nim]: https://github.com/nim-lang/Nim/issues/18095

File: beacon_chain/spec/forks_light_client.nim [hunk 3]
```diff
@@ -951,6 +1135,108 @@ func toDenebLightClientHeader(
...
+  # point to earlier forks. While Bellatrix blocks also contain an
+  # `ExecutionPayload` (minus `withdrawals_root`), it was not included
+  # in the corresponding light client data. To ensure compatibility
+  # with legacy data going through `upgrade_lc_header_to_capella`,
+  # leave out execution data.
+  electra.LightClientHeader(
+    beacon: blck.message.toBeaconBlockHeader())
+
+func toElectraLightClientHeader(
+    blck:  # `SomeSignedBeaconBlock` doesn't work here (Nim 1.6)
```
Comment (tersec) [inline beacon_chain/spec/forks_light_client.nim]: https://github.com/nim-lang/Nim/issues/18095

This way it won't be a mystery to check if/when it's eventually fixed

File: beacon_chain/spec/forks_light_client.nim [hunk 4]
```diff
@@ -951,6 +1135,108 @@ func toDenebLightClientHeader(
...
+      extra_data: payload.extra_data,
+      base_fee_per_gas: payload.base_fee_per_gas,
+      block_hash: payload.block_hash,
+      transactions_root: hash_tree_root(payload.transactions),
+      withdrawals_root: hash_tree_root(payload.withdrawals)),
+    execution_branch: blck.message.body.build_proof(
+      capella.EXECUTION_PAYLOAD_GINDEX).get)
+
+func toElectraLightClientHeader(
+    blck:  # `SomeSignedBeaconBlock` doesn't work here (Nim 1.6)
```
Comment (tersec) [inline beacon_chain/spec/forks_light_client.nim]: https://github.com/nim-lang/Nim/issues/18095

File: beacon_chain/spec/forks_light_client.nim [hunk 5]
```diff
@@ -951,6 +1135,108 @@ func toDenebLightClientHeader(
...
+      block_hash: payload.block_hash,
+      transactions_root: hash_tree_root(payload.transactions),
+      withdrawals_root: hash_tree_root(payload.withdrawals),
+      blob_gas_used: payload.blob_gas_used,
+      excess_blob_gas: payload.excess_blob_gas),
+    execution_branch: blck.message.body.build_proof(
+      capella.EXECUTION_PAYLOAD_GINDEX).get)
+
+func toElectraLightClientHeader(
+    blck:  # `SomeSignedBeaconBlock` doesn't work here (Nim 1.6)
```
Comment (tersec) [inline beacon_chain/spec/forks_light_client.nim]: https://github.com/nim-lang/Nim/issues/18095

File: beacon_chain/spec/forks_light_client.nim [hunk 6]
```diff
@@ -951,6 +1135,108 @@ func toDenebLightClientHeader(
...
+      excess_blob_gas: payload.excess_blob_gas,
+      deposit_requests_root: hash_tree_root(payload.deposit_requests),
+      withdrawal_requests_root: hash_tree_root(payload.withdrawal_requests),
+      consolidation_requests_root:
+        hash_tree_root(payload.consolidation_requests)),
+    execution_branch: blck.message.body.build_proof(
+      capella.EXECUTION_PAYLOAD_GINDEX).get)
+
 func toLightClientHeader*(
     blck:  # `SomeSignedBeaconBlock` doesn't work here (Nim 1.6)
```
Comment (tersec) [inline beacon_chain/spec/forks_light_client.nim]: https://github.com/nim-lang/Nim/issues/18095

File: beacon_chain/spec/forks_light_client.nim [hunk 7]
```diff
@@ -990,9 +1275,13 @@ func shortLog*[
       capellaData: typeof(x.capellaData.shortLog())
     of LightClientDataFork.Deneb:
       denebData: typeof(x.denebData.shortLog())
+    of LightClientDataFork.Electra:
+      electraData: typeof(x.electraData.shortLog())
 
   let xKind = x.kind  # Nim 1.6.12: Using `kind: x.kind` inside case is broken
```
Comment (tersec) [inline beacon_chain/spec/forks_light_client.nim]: Was this ever isolated/reproduced/fixed/reported/already-reported-upstream or maybe fixed in 2.0.6?

**Comment by tersec** [comment]
Also for all the "... works in Nim 1.6 ..." etc comments, as of the last couple days, `nimbus-eth2` uses Nim 2.0.6, so that's the relevant version to check. No need to find/update all of them, though, apparently the upstream issue isn't fixed, but at least new comments going in should reference 2.0 or 2.0.6 specifically.

File: beacon_chain/spec/datatypes/electra.nim [hunk 8]
```diff
@@ -675,6 +676,233 @@ func shortLog*(v: ExecutionPayload): auto =
...
+      return false
+
+  if epoch < cfg.DENEB_FORK_EPOCH:
+    if header.execution.blob_gas_used != 0 or
+        header.execution.excess_blob_gas != 0:
+      return false
+
+  if epoch < cfg.CAPELLA_FORK_EPOCH:
+    return
+      header.execution == default(ExecutionPayloadHeader) and
```
Comment (tersec) [inline beacon_chain/spec/datatypes/electra.nim]: In general,
```nim
type L = object
  a: string
  b: int
  c: array[32, byte]
  d: array[32, byte]
  e: uint64

func a(x: L): bool =
  x == default(L)

discard a(L())
```
produces, with Nim 2.0.6 using `refc`,
```c
struct tyObject_L__0ubpPjj6L2pwUQ1bt60yqg {
  NimStringDesc *a;
  NI b;
  tyArray__vEOa9c5qaE9ajWxR5R4zwfQg c;
  tyArray__vEOa9c5qaE9ajWxR5R4zwfQg d;
  NU64 e;
};

...

N_LIB_PRIVATE N_NIMCALL(NIM_BOOL,
                        a__a_u7)(tyObject_L__0ubpPjj6L2pwUQ1bt60yqg *x_p0) {
  NIM_BOOL result;
  tyObject_L__0ubpPjj6L2pwUQ1bt60yqg T1_;
  result = (NIM_BOOL)0;
  nimZeroMem((void *)(&T1_), sizeof(tyObject_L__0ubpPjj6L2pwUQ1bt60yqg));
  result = eqeq___a_u15(x_p0, (&T1_));
  popFrame();
  return result;
}
```

whereas
```nim
type L = object
  a: string
  b: int
  c: array[32, byte]
  d: array[32, byte]
  e: uint64

func a(x: L): bool =
  x == static(default(L))

discard a(L())
```
results in
```c
struct tyObject_L__0ubpPjj6L2pwUQ1bt60yqg {
  NimStringDesc *a;
  NI b;
  tyArray__vEOa9c5qaE9ajWxR5R4zwfQg c;
  tyArray__vEOa9c5qaE9ajWxR5R4zwfQg d;
  NU64 e;
};

...

static NIM_CONST tyObject_L__0ubpPjj6L2pwUQ1bt60yqg
    TM__R8RUzYq41iOx0I9bZH5Nyrw_2 = {
        ((NimStringDesc *)NIM_NIL),
        ((NI)0),
        {((NU8)0), ((NU8)0), ((NU8)0), ((NU8)0), ((NU8)0), ((NU8)0), ((NU8)0),
         ((NU8)0), ((NU8)0), ((NU8)0), ((NU8)0), ((NU8)0), ((NU8)0), ((NU8)0),
         ((NU8)0), ((NU8)0), ((NU8)0), ((NU8)0), ((NU8)0), ((NU8)0), ((NU8)0),
         ((NU8)0), ((NU8)0), ((NU8)0), ((NU8)0), ((NU8)0), ((NU8)0), ((NU8)0),
         ((NU8)0), ((NU8)0), ((NU8)0), ((NU8)0)},
        {((NU8)0), ((NU8)0), ((NU8)0), ((NU8)0), ((NU8)0), ((NU8)0), ((NU8)0),
         ((NU8)0), ((NU8)0), ((NU8)0), ((NU8)0), ((NU8)0), ((NU8)0), ((NU8)0),
         ((NU8)0), ((NU8)0), ((NU8)0), ((NU8)0), ((NU8)0), ((NU8)0), ((NU8)0),
         ((NU8)0), ((NU8)0), ((NU8)0), ((NU8)0), ((NU8)0), ((NU8)0), ((NU8)0),
         ((NU8)0), ((NU8)0), ((NU8)0), ((NU8)0)},
        0ULL};

...

N_LIB_PRIVATE N_NIMCALL(NIM_BOOL,
                        a__a_u7)(tyObject_L__0ubpPjj6L2pwUQ1bt60yqg *x_p0) {
  NIM_BOOL result;
  result = (NIM_BOOL)0;
  result = eqeq___a_u15(x_p0, (&TM__R8RUzYq41iOx0I9bZH5Nyrw_2));
  popFrame();
  return result;
}
```

It's possible gcc and clang can both under `-O3 -flto` deduce that they don't have to consume the stack space, reinitialize/zero every time, et cetera, but that seems fairly excessive when wrapping the `default(...)` as `static(default(...))` or using `const foo = default(...); compare_with(foo)` avoids the per-call construction/zeroing/stack space altogether of this object.

File: beacon_chain/spec/forks_light_client.nim (hunk 7 above)
Comment (etan-status) [reply beacon_chain/spec/forks_light_client.nim]: not reported upstream, but this is the error that is produced when trying to inline `x.kind`:

```
Error: a case selecting discriminator 'kind' with value 'x.kind' appears in the object construction, but the field(s) 'denebData' are in conflict with this value.
```

Still applies to 2.0.6, have updated the comment in https://github.com/status-im/nimbus-eth2/pull/6389

Comment (tersec) [reply beacon_chain/spec/forks_light_client.nim]: https://github.com/nim-lang/Nim/issues/23762

**Pull request merged**